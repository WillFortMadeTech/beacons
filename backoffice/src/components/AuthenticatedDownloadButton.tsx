import { Button } from "@mui/material";
import React, { MouseEventHandler } from "react";
import { useAuthContext } from "./auth/AuthProvider";

export function AuthenticatedDownloadButton({
  url,
}: {
  url: string;
}): JSX.Element | null {
  const link = React.useRef<HTMLAnchorElement>(null);

  const { user } = useAuthContext();

  const downloadFile =
    (accessToken: string | unknown): MouseEventHandler<HTMLAnchorElement> =>
    async () => {
      if (!link.current || link.current.href) {
        return;
      }

      const result = await fetch(url, {
        headers: { Authorization: `Bearer ${accessToken}` },
      });

      if (result.status === 503) {
        window.alert("There was an error while downloading.");
        return;
      }

      const filename = parseFilename(result.headers);

      if (!filename) {
        window.alert("There was an error while downloading.");
        return;
      }

      const blob = await result.blob();
      const href = window.URL.createObjectURL(blob);

      link.current.download = filename;
      link.current.href = href;

      link.current.click();
    };

  if (user.type !== "loggedInUser") {
    return null;
  }

  return (
    <Button
      component="a"
      ref={link}
      onClick={downloadFile(user.apiAccessToken)}
      color="inherit"
      variant="outlined"
      fullWidth
    >
      Export to Excel
    </Button>
  );
}

export const parseFilename = (headers: Headers): string | null => {
  const contentDisposition = headers.get("Content-Disposition");
  if (!contentDisposition) return null;

  return contentDisposition.split("filename=")[1];
};
