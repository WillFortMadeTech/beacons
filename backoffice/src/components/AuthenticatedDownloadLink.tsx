import { Button } from "@mui/material";
import React, { MouseEventHandler } from "react";
import { AuthContext } from "./auth/AuthWrapper";

export function AuthenticatedDownloadLink({
  url,
  filename,
}: {
  url: string;
  filename: string;
}): JSX.Element {
  const [loading, setLoading] = React.useState(false);
  const link = React.useRef<HTMLAnchorElement>(null);

  const handleLoading = () => {
    setLoading(true);
    setTimeout(() => {
      setLoading(false);
    }, 5000);
  };

  const downloadFile =
    (accessToken: string | unknown): MouseEventHandler<HTMLAnchorElement> =>
    async () => {
      if (!link.current || link.current.href) {
        return;
      }

      const result = await fetch(url, {
        headers: { Authorization: `Bearer ${accessToken}` },
      });

      if (result.status === 503) {
        handleLoading();
        return;
      }

      const blob = await result.blob();
      const href = window.URL.createObjectURL(blob);

      link.current.download = filename;
      link.current.href = href;

      link.current.click();
    };

  return (
    <>
      <AuthContext.Consumer>
        {(auth) => (
          <Button
            component="a"
            ref={link}
            onClick={downloadFile(auth.accessToken)}
            color="inherit"
            variant="outlined"
            fullWidth
            disabled={loading}
          >
            Export to Excel
          </Button>
        )}
      </AuthContext.Consumer>
    </>
  );
}
