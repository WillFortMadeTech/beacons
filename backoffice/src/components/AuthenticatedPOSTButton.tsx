import { Button } from "@mui/material";
import React, { MouseEventHandler } from "react";
import { logToServer } from "../utils/logger";
import { useAuthContext } from "./auth/AuthProvider";

/**
 * For triggering actions in the Service while authenticated as the user
 *
 * @param url the API endpoint to which to POST
 */
export function AuthenticatedPOSTButton({
  uri,
  children,
}: {
  uri: string;
  children: React.ReactNode;
}): JSX.Element | null {
  const { user } = useAuthContext();

  const performAction =
    (accessToken: string | unknown): MouseEventHandler<HTMLAnchorElement> =>
    async () => {
      await fetch(uri, {
        method: "POST",
        headers: { Authorization: `Bearer ${accessToken}` },
      }).catch((e) => {
        logToServer.error(`Error while POSTing to ${uri}: ${e}`);
      });
    };

  if (user.type !== "loggedInUser") {
    return null;
  }

  return (
    <Button
      component="a"
      onClick={performAction(user.apiAccessToken)}
      color="inherit"
      variant="outlined"
      fullWidth
    >
      {children}
    </Button>
  );
}
