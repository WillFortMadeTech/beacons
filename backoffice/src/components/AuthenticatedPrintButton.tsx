import { Button } from "@mui/material";
import React, { MouseEventHandler } from "react";
import { useAuthContext } from "./auth/AuthProvider";

export function AuthenticatedPrintButton({
  url,
  label,
  isFullWidth,
}: {
  url: string;
  label: string;
  isFullWidth: boolean;
}): JSX.Element | null {
  const link = React.useRef<HTMLAnchorElement>(null);

  const { user } = useAuthContext();

  const printFile =
    (accessToken: string | unknown): MouseEventHandler<HTMLAnchorElement> =>
    async () => {
      if (!link.current || link.current.href) {
        return;
      }

      // to-do: abstract into separate function to get the file
      const result = await fetch(url, {
        headers: { Authorization: `Bearer ${accessToken}` },
      });

      if (result.status === 503) {
        window.alert("There was an error while preparing the file.");
        return;
      }

      const filename = parseFilename(result.headers);

      if (!filename) {
        window.alert("There was an error while preparing the file.");
        return;
      }

      // then print it
      const file = await result.blob();
      const href = window.URL.createObjectURL(file);
      window.open(href, "PRINT", "height=400,width=600");
    };

  if (user.type !== "loggedInUser") {
    return null;
  }

  return (
    <Button
      component="a"
      ref={link}
      onClick={printFile(user.apiAccessToken)}
      color="inherit"
      variant="outlined"
      fullWidth={isFullWidth}
    >
      {label}
    </Button>
  );
}

export const parseFilename = (headers: Headers): string | null => {
  const contentDisposition = headers.get("Content-Disposition");
  if (!contentDisposition) {
    return null;
  }

  return contentDisposition.split("filename=")[1];
};
