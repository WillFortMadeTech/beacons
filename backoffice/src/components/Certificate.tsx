import { Theme } from "@mui/material/styles";
import createStyles from "@mui/styles/createStyles";
import makeStyles from "@mui/styles/makeStyles";
import React, { FunctionComponent } from "react";
import { Grid } from "@mui/material";
import { ICertificate } from "gateways/exports/ICertificate";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      fontFamily: "Arial",
      backgroundColor: "white",
      padding: "0.5%",
    },
    header: {
      borderBottom: "3px solid black",
    },
  })
);

interface CertificateProps {
  certificate: ICertificate;
}

export const Certificate: FunctionComponent<CertificateProps> = ({
  certificate,
}): JSX.Element => {
  const classes = useStyles();
  return (
    <Grid container className={classes.root}>
      <Grid className={classes.header} container spacing={0.5}>
        <Grid item xl={12}>
          <div>OFFICIAL</div>
          <h3>UK Distress & Security Beacon Registration</h3>
        </Grid>
        <Grid item xs={6}>
          <div>
            PROOF OF REGISTRATION: {certificate.proofOfRegistrationDate}
          </div>
          <Grid item xs={6}>
            <div>DEPT REF: {certificate.beacon.referenceNumber}</div>
          </Grid>
        </Grid>
      </Grid>

      <Grid container spacing={0.5}>
        <Grid item xs={6}>
          <div>RECORD CREATED DATE: {certificate.beacon.registeredDate}</div>
        </Grid>
        <Grid item xs={6}>
          <div>BEACON STATUS: {certificate.beacon.status}</div>
        </Grid>
        <Grid item xs={6}>
          <div>LAST MODIFIED: {certificate.beacon.lastModifiedDate}</div>
        </Grid>
      </Grid>

      {/* // beacon details */}

      <Grid item xs={6}>
        <div>MCA's Contact Number: {certificate.mcaContactNumber}</div>
      </Grid>

      <Grid item xs={6}>
        <div>Csta: {certificate.beacon.csta}</div>
      </Grid>
      <Grid item xs={6}>
        <div>Hex ID: {certificate.beacon.hexId}</div>
      </Grid>
      <Grid item xs={6}>
        <div>Last modified: {certificate.beacon.lastModifiedDate}</div>
      </Grid>
      <Grid item xs={6}>
        <div>MTI: {certificate.beacon.mti}</div>
      </Grid>
      <Grid item xs={6}>
        <div>Vessel Name: {certificate.uses[0].vesselName}</div>
      </Grid>
      <Grid item xs={6}>
        <div>Protocol: {certificate.beacon.protocol}</div>
      </Grid>
      <Grid item xs={6}>
        <div>SVDR: {certificate.beacon.svdr}</div>
      </Grid>
    </Grid>
  );
};
