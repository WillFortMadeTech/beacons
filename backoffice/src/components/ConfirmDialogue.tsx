import {
  Button,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from "@mui/material";
import { FunctionComponent } from "react";

interface IConfirmDialogueProps {
  confirmDialogueModel: IConfirmDialogueModel;
  confirmOption: (actionOptionSelected: boolean) => void;
}

export interface IConfirmDialogueModel {
  dialogueTitle: string;
  dialogueContentText: string;
  action: string;
  dismissal: string;
}

export const ConfirmDialogue: FunctionComponent<IConfirmDialogueProps> = ({
  confirmDialogueModel,
  confirmOption,
}): JSX.Element => {
  function handleClick(isActionOption: boolean): void {
    confirmOption(isActionOption);
  }

  return (
    <div>
      <DialogTitle id="alert-dialog-title">
        {confirmDialogueModel.dialogueTitle}
      </DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">
          {confirmDialogueModel.dialogueContentText}
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => handleClick(false)}>
          {confirmDialogueModel.dismissal}
        </Button>
        <Button onClick={() => handleClick(true)} autoFocus>
          {confirmDialogueModel.action}
        </Button>
      </DialogActions>
    </div>
  );
};
