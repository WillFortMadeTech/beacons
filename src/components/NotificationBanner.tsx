import React, { FunctionComponent, ReactNode } from "react";

interface NotificationBannerProps {
  title: string;
  children: ReactNode;
}

export const NotificationBanner: FunctionComponent<NotificationBannerProps> = ({
  title,
  children,
}: NotificationBannerProps): JSX.Element => (
  <div
    className="govuk-notification-banner"
    role="region"
    aria-labelledby="govuk-notification-banner-title"
    data-module="govuk-notification-banner"
  >
    <div className="govuk-notification-banner__header">
      <h2
        className="govuk-notification-banner__title"
        id="govuk-notification-banner-title"
      >
        {title}
      </h2>
    </div>
    <div className="govuk-notification-banner__content">
      <div className="govuk-notification-banner__heading">{children}</div>
    </div>
  </div>
);

export const NotificationBannerSuccess: FunctionComponent<NotificationBannerProps> = ({
  title,
  children,
}): JSX.Element => (
  <div
    className="govuk-notification-banner govuk-notification-banner--success"
    role="alert"
    aria-labelledby="govuk-notification-banner-title"
    data-module="govuk-notification-banner"
  >
    <div className="govuk-notification-banner__header">
      <h2
        className="govuk-notification-banner__title"
        id="govuk-notification-banner-title"
      >
        {title}
      </h2>
    </div>
    <div className="govuk-notification-banner__content">{children}</div>
  </div>
);
