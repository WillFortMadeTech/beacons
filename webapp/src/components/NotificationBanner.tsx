import React, { FunctionComponent, ReactNode } from "react";

export interface NotificationBannerProps {
  isErrorMessage: boolean;
  title: string;
  heading: string;
  children: ReactNode;
}

interface NotificationBannerSuccessProps {
  title: string;
  children: ReactNode;
}

export const NotificationBanner: FunctionComponent<NotificationBannerProps> = (
  notificationBannerProps: NotificationBannerProps
): JSX.Element => (
  <div
    className={
      notificationBannerProps.isErrorMessage
        ? "govuk-notification-banner govuk-error-notification-banner"
        : "govuk-notification-banner"
    }
    role="region"
    aria-labelledby="govuk-notification-banner-title"
    data-module="govuk-notification-banner"
  >
    <div className={"govuk-notification-banner__header"}>
      <h2
        className="govuk-notification-banner__title"
        id="govuk-notification-banner-title"
      >
        {notificationBannerProps.title}
      </h2>
    </div>
    <div className="govuk-notification-banner__content">
      <p className="govuk-notification-banner__heading">
        {notificationBannerProps.heading}
      </p>
      {notificationBannerProps.children}
    </div>
  </div>
);

export const NotificationBannerSuccess: FunctionComponent<
  NotificationBannerSuccessProps
> = ({ title, children }: NotificationBannerSuccessProps): JSX.Element => (
  <div
    className="govuk-notification-banner govuk-notification-banner--success"
    role="alert"
    aria-labelledby="govuk-notification-banner-title"
    data-module="govuk-notification-banner"
  >
    <div className="govuk-notification-banner__header">
      <h2
        className="govuk-notification-banner__title"
        id="govuk-notification-banner-title"
      >
        {title}
      </h2>
    </div>
    <div className="govuk-notification-banner__content">{children}</div>
  </div>
);
