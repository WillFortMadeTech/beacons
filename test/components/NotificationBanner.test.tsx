import { render } from "@testing-library/react";
import React from "react";
import { NotificationBanner } from "../../src/components/NotificationBanner";

describe("Notification Banner component", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <NotificationBanner title="test001">Inner message</NotificationBanner>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
