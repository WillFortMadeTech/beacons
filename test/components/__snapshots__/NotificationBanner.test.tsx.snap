// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Banner component renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="govuk-notification-banner-title"
    class="govuk-notification-banner"
    data-module="govuk-notification-banner"
    role="region"
  >
    <div
      class="govuk-notification-banner__header"
    >
      <h2
        class="govuk-notification-banner__title"
        id="govuk-notification-banner-title"
      >
        test001
      </h2>
    </div>
    <div
      class="govuk-notification-banner__content"
    >
      <div
        class="govuk-notification-banner__heading"
      >
        Inner message
      </div>
    </div>
  </div>
</DocumentFragment>
`;
